third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -197,6 +197,17 @@ impl ControlFlowGraphSimplifierTransformation {
 /// 1. construct a `ControlFlowGraphCodeGenerator` from the code,
 /// 2. perform the transformation on the `ControlFlowGraphCodeGenerator`
 /// 3. generate the code back from the `ControlFlowGraphCodeGenerator`
+///
+/// Invariants:
+/// 1. Fields `code_blocks`, `successors`, and `predecessors` have the same set of keys.
+/// 2. Block `A` has sucessor block `B` iff block `B` has predecessor block `A`, according
+/// to fields `successors` and `predecessors` respectively.
+/// 3. In `code_blocks`, entry and exit blocks cannot have code (mapped to empty vectors in `code_blocks`),
+/// and all other blocks must have code.
+/// 4. Entry block has exactly one successor.
+/// 5. All blocks except for the exit block must have at least one successor.
+/// 6. All blocks with predecessors must start with a label, unless it's the first block, or the exit block.
```

Comment:
Yes, thus the "unless" part.

Reply:
As written: unless a block is the first block, .. => first block, even if it has predecessors, need not start with a label (it is excluded because of the unless).

But this is not true. First block, when it has a predecessor other than the entry, must have a label.

Reply:
I see. Corrected.